src/java/com/unifina/datasource/DataSource.java

```diff
@@ -83,69 +98,103 @@ public void addStopListener(IStopListener stopListener) {
 		stopListeners.add(stopListener);
 	}
 
-	public void startFeed() {
+	/**
+	 * Starts the main event loop and blocks until it is stopped.
+	 * This should be called from the event processing Thread.
+	 */
+	public void start() {
 		for (IStartListener startListener : startListeners) {
 			startListener.onStart();
 		}
 
+		started = true;
+
+		// Collect all the required StreamPartitions
+		Set<StreamPartition> allStreamPartitions = new HashSet<>();
+		eventRecipientsByStreamPartitions.keySet().forEach((streamPartitions -> allStreamPartitions.addAll(streamPartitions)));
+
 		try {
-			started = true;
-			doStartFeed();
+			streamMessageSource = createStreamMessageSource(
+				allStreamPartitions,
+				new StreamMessageSource.StreamMessageConsumer() {
+					@Override
+					public void accept(StreamMessage streamMessage) {
+						// Consult the router to find consumers who need to receive this StreamMessage
+						router.route(streamMessage)
+							.forEach(routedConsumer ->
+								// Enqueue an event for each consumer registered with the router
+								DataSource.this.accept(new Event<>(
+									streamMessage,
+									streamMessage.getTimestampAsDate(),
+									streamMessage.getSequenceNumber(),
+									routedConsumer
+								)));
+					}
+
+					@Override
+					public void done() {
+						// Enqueue an end event, which when processed, aborts the event queue.
+						DataSource.this.accept(new Event<>(
+							null,
+							globals.getEndDate() != null ? globals.getEndDate() : new Date(),
+							(nul) -> eventQueue.abort()
```

Comment:
The method is documented in `StreamMessageSource`:

```
	public abstract static class StreamMessageConsumer implements Consumer<StreamMessage> {
		/**
		 * Signals that there will be no further messages from the message source.
		 */
		public abstract void done();
	}
```

I believe it's used in historical mode to signal end of messages. It would indeed be more elegant to call `DataSource#abort()` instead of `eventQueue.abort()`. I'll try it out.